### ISSUE TITLE
Problem saving ontology in Turtle format - owlapi 4

### ISSUE THREAD

**Issue by lpalaciosmedinacelli** [owlcs/owlapi#905]
Hi, 

I have been struggling for a while with an issue when saving an ontology into Turtle format, and I can't seem to pinpoint the problem.

I basically need to pass the onotlogy to a reasoner that I am testing, but the reasoner accepts only ontologies in Turtle format.

My code goes like this, where oALL is my OWLOntology:

```
TurtleDocumentFormat turtleFormat = new TurtleDocumentFormat();
ByteArrayOutputStream os= new ByteArrayOutputStream ();         
manager.saveOntology(oALL,turtleFormat,os);

```
Here 'os' is a ByteArrayOutputStream that should end containing the Turtle ontology, so I can pass it to the reasoner. But when I execute my code I get the following error:

```
org.semanticweb.owlapi.model.OWLOntologyStorageException: org.openrdf.rio.UnsupportedRDFormatException: No writer factory available for RDF format Turtle (mimeTypes=text/turtle, application/x-turtle; ext=ttl)
	at org.semanticweb.owlapi.rio.RioStorer.getRDFHandlerForWriter(RioStorer.java:139)
	at org.semanticweb.owlapi.rio.RioStorer.storeOntology(RioStorer.java:244)
	at org.semanticweb.owlapi.util.AbstractOWLStorer.storeOntology(AbstractOWLStorer.java:119)
	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.saveOntology(OWLOntologyManagerImpl.java:1310)
	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.saveOntology(OWLOntologyManagerImpl.java:1287)
	at consult.askOntologyJRDF.getTypes(askOntologyJRDF.java:461)
	at learning.UnsupervisedLearning.learnSample(UnsupervisedLearning.java:191)
	at main.MainAccess.sendFeedback(MainAccess.java:1499)
	at test.TestFeedback.optionTwo(TestFeedback.java:185)
	at test.TestFeedback.main(TestFeedback.java:31)
Caused by: org.openrdf.rio.UnsupportedRDFormatException: No writer factory available for RDF format Turtle (mimeTypes=text/turtle, application/x-turtle; ext=ttl)
	at org.openrdf.rio.Rio.createWriter(Rio.java:257)
	at org.semanticweb.owlapi.rio.RioStorer.getRDFHandlerForWriter(RioStorer.java:137)
	... 9 more
```

At some point I could bypass the problem usin RIO Turtle format like this:

`RioTurtleDocumentFormat rdfTurtleFormat = new RioTurtleDocumentFormat();`

Which worked locally in my Mac, but when I upload the code to the Linux server, the above problem rises up. Any hints? I have not been able to find a solution. Help is appreciated.

**Comment by ignazio1977** [comment]
Looks like missing dependencies on the classpath when executing on your server. How do you build your application?

**Comment by lpalaciosmedinacelli** [comment]
Hi there!

I use eclipse Neon, and I build a Runnable jar file, where librabries and dependencies should be packed with the deployable jar (right?). OWLApi in particular is a maven dependency (4.2.3).

**Comment by ignazio1977** [comment]
Some dependencies are getting packed (the stack trace comes from OWLAPI classes so they are present) however the list of parsers and storers is included as meta-inf/services files and those might not have been added as expected - it certainly would explain the exception, and also the fact that rio turtle seems to work sometimes.

Which artifact do you have in your pom file? Is it `apibinding`, `distribution` or `osgidistribution`?

If you open the runnable jar and attach here the files in the meta-inf folder, that would help as well.

**Comment by lpalaciosmedinacelli** [comment]
Hello Ignazio, 

As always, tanks for the reply. Yes, some dependencies are there, that's why it was difficult to detect whats the problem is, I was not aware of the meta-inf/services, makes sense.

OWLApi, is included in the POM as follows:
```
<dependency>
    <groupId>net.sourceforge.owlapi</groupId>
    <artifactId>owlapi-distribution</artifactId>
    <version>4.2.3</version>
</dependency>
```
As for the meta-inf folder in the .jar  there is one file:
`META-INF/MANIFEST.MF`
whose contents are:

```
Manifest-Version: 1.0
Rsrc-Class-Path: ./ junit-4.4.jar owlapi-distribution-4.2.3
 .jar jackson-core-2.5.1.jar jackson-databind-2.5.1.jar jackson-annota
 tions-2.5.1.jar xz-1.5.jar sesame-model-2.7.16.jar sesame-util-2.7.16
 .jar sesame-rio-api-2.7.16.jar sesame-rio-languages-2.7.16.jar sesame
 -rio-datatypes-2.7.16.jar sesame-rio-binary-2.7.16.jar sesame-rio-n3-
 2.7.16.jar sesame-rio-nquads-2.7.16.jar sesame-rio-ntriples-2.7.16.ja
 r sesame-rio-rdfjson-2.7.16.jar sesame-rio-rdfxml-2.7.16.jar sesame-r
 io-trix-2.7.16.jar sesame-rio-turtle-2.7.16.jar sesame-rio-trig-2.7.1
 6.jar jsonld-java-sesame-0.5.0.jar jsonld-java-0.5.0.jar httpclient-c
 ache-4.2.5.jar httpclient-4.2.5.jar httpcore-4.2.4.jar jcl-over-slf4j
 -1.7.7.jar semargl-sesame-0.6.1.jar semargl-core-0.6.1.jar semargl-rd
 fa-0.6.1.jar semargl-rdf-0.6.1.jar trove4j-3.0.3.jar guava-18.0.jar g
 uice-4.0.jar javax.inject-1.jar aopalliance-1.0.jar guice-assistedinj
 ect-4.0.jar guice-multibindings-4.0.jar jsr305-2.0.1.jar slf4j-api-1.
 7.10.jar commons-io-2.4.jar rdf4j-rio-turtle-2.4.0.jar rdf4j-model-2.
 4.0.jar rdf4j-rio-api-2.4.0.jar rdf4j-rio-datatypes-2.4.0.jar rdf4j-r
 io-languages-2.4.0.jar rdf4j-util-2.4.0.jar slf4j-simple-1.7.21.jar p
 oi-3.15.jar commons-codec-1.10.jar commons-collections4-4.1.jar poi-o
 oxml-3.15.jar poi-ooxml-schemas-3.15.jar xmlbeans-2.6.0.jar stax-api-
 1.0.1.jar curvesapi-1.04.jar org.semanticweb.hermit-1.3.8.413.jar com
 mons-logging-1.1.3.jar axiom-api-1.2.14.jar geronimo-activation_1.1_s
 pec-1.1.jar geronimo-javamail_1.4_spec-1.7.1.jar jaxen-1.1.4.jar gero
 nimo-stax-api_1.0_spec-1.0.1.jar apache-mime4j-core-0.7.2.jar axiom-c
 14n-1.2.14.jar axiom-impl-1.2.14.jar woodstox-core-asl-4.1.4.jar stax
 2-api-3.1.1.jar axiom-dom-1.2.14.jar automaton-1.11-8.jar jgrapht-cor
 e-1.0.1.jar json-20170516.jar json-simple-1.1.1.jar
Class-Path: .
Rsrc-Main-Class: test.TestFeedback
Main-Class: org.eclipse.jdt.internal.jarinjarloader.JarRsrcLoader
```

Quite interesting to see several of my dependencies here, but I dont see they  (explicitly) mention TurtleDocumentFormat.

Does this shade some light on the issue?

**Comment by ignazio1977** [comment]
Yes thank you. owlapi-distribution.jar should contain the files you need but it seems they're not available to ServiceLoader as expected. Possibly just a matter of not being in the right classpath manager.

One possible workaround: open owlapi-distribution and copy the meta-inf/services files to your project (same folder structure) so that they're in you runnable jar, and see if that makes a difference.

**Comment by lpalaciosmedinacelli** [comment]
Hi!

As you suggested, I opened the owlapi distribution `owlapi-distribution-4.2.3.jar` and I saw all packages and dependencies are there in the right place, as well as the meta-inf/services folder. I was not sure how to add these to my .jar/project (I was trying to open the .jar and just add the /services in the meta-inf/ folder, but then, should I also modify the MANIFEST.MF for it to work?) Since in MY jar, owlapi was also packed,  I was not sure how to proceed, and I didn't like too much the idea of modifying manually the .jar since I might easily ruin it.

But you correctly pointed out the problem: dependencies are just not being loaded. So I exported my runnable .jar with the option : **"Copy required libraries into a sub-folder...".** So that owlapi would be complete. And it works!!! No errors so far!! (btw. I returned to use `RioTurtleDocumentFormat`)

So first of all, thank you! I couldn't have made it work without your help, and thanks for answering on the weekend, we appreciate that!

Now, I don't like to have a .jar and an additional folder woth the libraries, things should work properly, so if you have any suggestions/info or comments on how to properly add those dependencies, (the manual process you described before) will help.

Strange that the export method changes the outcome, isn't it?
